File: go/terminal/ui/textfield.go
package ui

import (
	"strings"
	"time"
	"unicode"

	"github.com/nullentropy/caution/go/terminal/gfx"
	"github.com/nullentropy/caution/go/terminal/text"
)

type TextField struct {
	Core
	Value       string
	Placeholder string
	// Radius overrides the radius.control metric for this field (nil = the
	// token): pill search boxes and other chrome-strip shapes.
	Radius *float32
	// FontSize in logical px (`size` prop) and family (`mono` prop).
	FontSize float32
	Mono     bool
	// Focused mirrors Ui focus so the protocol layer can honor local echo
	// (a focused field ignores server value sets).
	Focused bool

	// OnInput fires on every edit (wire-level debouncing happens in the
	// protocol layer), and OnCommit on Enter/blur, only when changed.
	OnInput  func(v string)
	OnCommit func(v string)

	// Last-seen values of the server's one-shot command props (FocusSeq
	// lives on Core, since it's universal).
	ResetSeq    float32
	OverrideSeq float32

	// selEnd is the caret, and selStart pins the other end of the selection.
	selStart, selEnd int
	anchor           int
	scrollX          float32
	lastCommitted    string
	// lastActivity anchors the blink phase: the caret is always solid right
	// after it moves, which needs no per-field timer.
	lastActivity time.Time

	// Snapshot undo: each entry is the state before a mutating edit. Runs of
	// single-rune typing coalesce into one entry, like every editor.
	undo, redo []tfSnapshot
	lastTyping time.Time
}

type tfSnapshot struct {
	value            string
	selStart, selEnd int
}

const tfUndoCap = 200

const tfBlink = 530 * time.Millisecond

// tfPadX is the field's inner text inset: the space.pad token.
func tfPadX() float32 { return Metrics.SpacePad }

func NewTextField() *TextField {
	t := &TextField{FontSize: 14}
	t.self = t
	return t
}

func (t *TextField) font() gfx.Font {
	return gfx.NewFont(t.FontSize, gfx.FontOpts{Mono: t.Mono})
}

func (t *TextField) IntrinsicSize() Size {
	h := max(Metrics.ControlHeight, float32(int(t.FontSize*1.7+0.5)))
	return Size{220, h}
}

func (t *TextField) Interactive() bool { return true }
func (t *TextField) Focusable() bool   { return true }
func (t *TextField) Cursor() string    { return "text" }

func (t *TextField) touch() { t.lastActivity = time.Now() }

func (t *TextField) selRange() (int, int) {
	if t.selStart <= t.selEnd {
		return t.selStart, t.selEnd
	}
	return t.selEnd, t.selStart
}

// Commit fires OnCommit if the value changed since focus/last commit.
func (t *TextField) Commit() {
	if t.Value == t.lastCommitted {
		return
	}
	t.lastCommitted = t.Value
	if t.OnCommit != nil {
		t.OnCommit(t.Value)
	}
}

// ForceClear is the server-forced clear (`ClearValue` in the SDK). Unlike a
// plain value set it applies even while focused, for submit-and-keep-typing
// flows.
func (t *TextField) ForceClear() {
	t.Value = ""
	t.lastCommitted = ""
	t.selStart, t.selEnd, t.anchor = 0, 0, 0
	t.scrollX = 0
	t.undo, t.redo = nil, nil // a server-forced transition starts fresh history
	t.lastTyping = time.Time{}
	t.touch()
	t.Invalidate()
}

// ForceValue is the server-forced value (`SetValueNow` in the SDK). It applies
// even while focused, for validation and formatting. The pushed value becomes
// the new baseline, so Escape reverts to it and no commit fires unless the user
// edits again. History restarts from it.
func (t *TextField) ForceValue(v string) {
	t.Value = v
	t.lastCommitted = v
	n := len([]rune(v))
	t.selStart, t.selEnd, t.anchor = n, n, n
	t.undo, t.redo = nil, nil
	t.lastTyping = time.Time{}
	t.touch()
	t.Invalidate()
}

func (t *TextField) SelectAll() {
	t.selStart, t.anchor = 0, 0
	t.selEnd = len([]rune(t.Value))
	t.touch()
	t.Invalidate()
}

func (t *TextField) OnFocusChange(focused bool) {
	t.Focused = focused
	if focused {
		t.lastCommitted = t.Value
		t.touch()
	} else {
		t.Commit()
	}
	t.Invalidate()
}

// -- editing operations ---------------------------------------------------------

// edit replaces rune range [s, e) with insert and puts the caret after it.
func (t *TextField) edit(s, e int, insert string) {
	t.editKind(s, e, insert, false)
}

// editKind is edit with a coalescing hint: consecutive typing edits share
// one undo entry.
func (t *TextField) editKind(s, e int, insert string, typing bool) {
	rs := []rune(t.Value)
	s = clampInt(s, 0, len(rs))
	e = clampInt(e, s, len(rs))
	next := string(rs[:s]) + insert + string(rs[e:])
	caret := s + len([]rune(insert))
	changed := next != t.Value
	if changed {
		t.pushUndo(typing)
	}
	t.Value = next
	t.selStart, t.selEnd, t.anchor = caret, caret, caret
	t.touch()
	if changed && t.OnInput != nil {
		t.OnInput(next)
	}
	t.Invalidate()
}

// -- undo ---------------------------------------------------------------------

func (t *TextField) pushUndo(typing bool) {
	now := time.Now()
	coalesce := typing && !t.lastTyping.IsZero() && now.Sub(t.lastTyping) < time.Second && len(t.undo) > 0
	if typing {
		t.lastTyping = now
	} else {
		t.lastTyping = time.Time{}
	}
	t.redo = nil // a fresh edit invalidates the redo branch
	if coalesce {
		return // the run's first snapshot already holds the pre-typing state
	}
	t.undo = append(t.undo, tfSnapshot{t.Value, t.selStart, t.selEnd})
	if len(t.undo) > tfUndoCap {
		t.undo = t.undo[len(t.undo)-tfUndoCap:]
	}
}

func (t *TextField) restore(s tfSnapshot) {
	changed := s.value != t.Value
	t.Value = s.value
	t.selStart, t.selEnd, t.anchor = s.selStart, s.selEnd, s.selEnd
	t.lastTyping = time.Time{}
	t.touch()
	if changed && t.OnInput != nil {
		t.OnInput(t.Value)
	}
	t.Invalidate()
}

func (t *TextField) undoEdit() {
	if len(t.undo) == 0 {
		return
	}
	top := t.undo[len(t.undo)-1]
	t.undo = t.undo[:len(t.undo)-1]
	t.redo = append(t.redo, tfSnapshot{t.Value, t.selStart, t.selEnd})
	t.restore(top)
}

func (t *TextField) redoEdit() {
	if len(t.redo) == 0 {
		return
	}
	top := t.redo[len(t.redo)-1]
	t.redo = t.redo[:len(t.redo)-1]
	t.undo = append(t.undo, tfSnapshot{t.Value, t.selStart, t.selEnd})
	t.restore(top)
}

// revert is escape-to-revert: back to the value the field had at focus (or
// last commit). Reported so the caller can decide whether Escape was
// consumed. A clean field lets it bubble (dialogs, selection clearing).
func (t *TextField) revert() bool {
	if t.Value == t.lastCommitted {
		return false
	}
	t.pushUndo(false) // revert itself is undoable
	t.Value = t.lastCommitted
	n := len([]rune(t.Value))
	t.selStart, t.selEnd, t.anchor = n, n, n
	t.touch()
	if t.OnInput != nil {
		t.OnInput(t.Value)
	}
	t.Invalidate()
	return true
}

func (t *TextField) moveCaret(to int, extend bool) {
	to = clampInt(to, 0, len([]rune(t.Value)))
	if extend {
		t.selEnd = to
	} else {
		t.selStart, t.selEnd, t.anchor = to, to, to
	}
	t.touch()
	t.Invalidate()
}

func isWordRuneTF(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsNumber(r) || r == '_'
}

func (t *TextField) wordLeft(from int) int {
	rs := []rune(t.Value)
	i := clampInt(from, 0, len(rs))
	for i > 0 && !isWordRuneTF(rs[i-1]) {
		i--
	}
	for i > 0 && isWordRuneTF(rs[i-1]) {
		i--
	}
	return i
}

func (t *TextField) wordRight(from int) int {
	rs := []rune(t.Value)
	i := clampInt(from, 0, len(rs))
	for i < len(rs) && !isWordRuneTF(rs[i]) {
		i++
	}
	for i < len(rs) && isWordRuneTF(rs[i]) {
		i++
	}
	return i
}

func (t *TextField) wordRangeAt(i int) (int, int) {
	rs := []rune(t.Value)
	if len(rs) == 0 {
		return 0, 0
	}
	i = clampInt(i, 0, len(rs)-1)
	target := isWordRuneTF(rs[i])
	s, e := i, i+1
	for s > 0 && isWordRuneTF(rs[s-1]) == target {
		s--
	}
	for e < len(rs) && isWordRuneTF(rs[e]) == target {
		e++
	}
	return s, e
}

// OnChar inserts printable input (the shell routes GLFW's char callback
// here). Control runes are dropped, and shortcuts arrive via OnKey instead.
func (t *TextField) OnChar(r rune) bool {
	if r < 0x20 || r == 0x7f {
		return false
	}
	t.sound("type")
	s, e := t.selRange()
	t.editKind(s, e, string(r), true)
	return true
}

func (t *TextField) OnKey(k Key) bool {
	n := len([]rune(t.Value))
	s, e := t.selRange()
	cmdish := k.Meta || k.Ctrl
	switch k.Name {
	case "Enter":
		t.Commit()
		return true
	case "Backspace":
		switch {
		case s != e:
			t.edit(s, e, "")
		case k.Meta:
			t.edit(0, t.selEnd, "") // Cmd+Backspace deletes to line start
		case k.Alt:
			t.edit(t.wordLeft(t.selEnd), t.selEnd, "") // Alt+Backspace deletes a word
		case t.selEnd > 0:
			t.edit(t.selEnd-1, t.selEnd, "")
		}
		return true
	case "Delete":
		if s != e {
			t.edit(s, e, "")
		} else if t.selEnd < n {
			t.edit(t.selEnd, t.selEnd+1, "")
		}
		return true
	case "Left":
		to := t.selEnd - 1
		switch {
		case k.Meta:
			to = 0
		case k.Alt:
			to = t.wordLeft(t.selEnd)
		case s != e && !k.Shift:
			to = s // a plain arrow collapses the selection to its edge
		}
		t.moveCaret(to, k.Shift)
		return true
	case "Right":
		to := t.selEnd + 1
		switch {
		case k.Meta:
			to = n
		case k.Alt:
			to = t.wordRight(t.selEnd)
		case s != e && !k.Shift:
			to = e
		}
		t.moveCaret(to, k.Shift)
		return true
	case "Up", "Home":
		t.moveCaret(0, k.Shift)
		return true
	case "Down", "End":
		t.moveCaret(n, k.Shift)
		return true
	case "a":
		if cmdish {
			t.SelectAll()
			return true
		}
	case "c":
		if cmdish && s != e {
			t.writeClipboard(s, e)
			return true
		}
	case "x":
		if cmdish {
			if s != e {
				t.writeClipboard(s, e)
				t.edit(s, e, "")
			}
			return true
		}
	case "v":
		if cmdish {
			t.paste()
			return true
		}
	case "z":
		if cmdish {
			if k.Shift {
				t.redoEdit()
			} else {
				t.undoEdit()
			}
			return true
		}
	case "Escape":
		return t.revert() // clean field: let Escape bubble (dialogs, selection)
	}
	return false
}

func (t *TextField) writeClipboard(s, e int) {
	if t.UI.WriteClipboard == nil {
		return
	}
	rs := []rune(t.Value)
	t.UI.WriteClipboard(string(rs[s:e]))
}

func (t *TextField) paste() {
	if t.UI.ReadClipboard == nil {
		return
	}
	// Single-line field: newlines flatten to spaces, other control runes drop.
	clip := strings.Map(func(r rune) rune {
		switch {
		case r == '\n' || r == '\t':
			return ' '
		case r < 0x20 || r == 0x7f:
			return -1
		}
		return r
	}, t.UI.ReadClipboard())
	if clip == "" {
		return
	}
	s, e := t.selRange()
	t.edit(s, e, clip)
}

// -- geometry ---------------------------------------------------------------------

// runeBoundaryXs maps every rune boundary of a shaped run (runeCount runes)
// to an x offset from the run origin, via shaping clusters (runes swallowed
// into a cluster sit at its start).
func runeBoundaryXs(run *text.Run, runeCount int) []float32 {
	xs := make([]float32, runeCount+1)
	seen := make([]bool, runeCount+1)
	for _, g := range run.Glyphs {
		if g.Cluster >= 0 && g.Cluster <= runeCount && !seen[g.Cluster] {
			xs[g.Cluster] = g.X
			seen[g.Cluster] = true
		}
	}
	xs[runeCount] = run.Width
	for i := 1; i < runeCount; i++ {
		if !seen[i] {
			xs[i] = xs[i-1]
		}
	}
	return xs
}

// boundaryXs maps every rune boundary of the value to an x offset from the
// text origin.
func (t *TextField) boundaryXs() []float32 {
	run := t.UI.Measure(t.font(), t.Value)
	return runeBoundaryXs(run, len([]rune(t.Value)))
}

func (t *TextField) xAt(xs []float32, i int) float32 {
	return xs[clampInt(i, 0, len(xs)-1)]
}

// runeAtX is the nearest rune boundary to an x offset from the text origin.
func (t *TextField) runeAtX(xs []float32, x float32) int {
	best, bestD := 0, float32(1e30)
	for i, bx := range xs {
		if d := abs32(x - bx); d < bestD {
			bestD, best = d, i
		}
	}
	return best
}

// -- painting -----------------------------------------------------------------------

func (t *TextField) PaintSelf(dl *gfx.DisplayList) {
	b := t.Bounds
	if t.UI.ShowFocusRing(t) {
		PaintFocusRing(dl, b, radiusOr(t.Radius, Metrics.RadiusControl))
	}
	border := Theme["controlEdge"]
	bw := Metrics.BorderWidth
	if t.Focused {
		border = Theme["accent"]
		bw = Metrics.BorderWidth * 1.5
	}
	dl.Rect(b, *Theme["panelInset"], gfx.RectOpts{
		Radius: gfx.CornerRadius(radiusOr(t.Radius, Metrics.RadiusControl)), BorderWidth: bw, BorderColor: border,
	})

	run := t.UI.Measure(t.font(), t.Value)
	xs := t.boundaryXs()
	innerW := b.W - tfPadX()*2
	textH := run.Ascent + run.Descent
	textY := b.Y + (b.H-textH)/2

	// Keep the caret inside the viewport by sliding the text.
	if t.Focused {
		caretX := t.xAt(xs, t.selEnd)
		if caretX-t.scrollX > innerW {
			t.scrollX = caretX - innerW
		}
		if caretX-t.scrollX < 0 {
			t.scrollX = caretX
		}
	}
	textX := b.X + tfPadX() - t.scrollX

	dl.PushClip(gfx.Inset(b, 1.5))
	if t.Value == "" && t.Placeholder != "" {
		dl.Text(t.Placeholder, b.X+tfPadX(), textY, t.font(), *Theme["inkFaint"])
	}
	s, e := t.selRange()
	if t.Focused && s != e {
		x0 := t.xAt(xs, s)
		x1 := t.xAt(xs, e)
		dl.Fill(gfx.R(textX+x0-1, textY-1, x1-x0+2, textH+2),
			gfx.WithAlpha(*Theme["accent"], 0.35), gfx.CornerRadius(2))
	}
	dl.Text(t.Value, textX, textY, t.font(), *Theme["ink"])
	if t.Focused && s == e && t.blinkOn() {
		dl.Fill(gfx.R(textX+t.xAt(xs, t.selEnd), textY-1, 1.5, textH+2),
			*Theme["accent"], gfx.Corners{})
	}
	dl.PopClip()
}

func (t *TextField) blinkOn() bool {
	return (time.Since(t.lastActivity)/tfBlink)%2 == 0
}

// blinkTick is how long until the caret phase next flips (the shell's wait
// timeout while a collapsed-caret field is focused).
func (t *TextField) blinkTick() time.Duration {
	return tfBlink - time.Since(t.lastActivity)%tfBlink
}

// -- mouse --------------------------------------------------------------------------

func (t *TextField) OnPointerDown(x, _ float32, detail int) {
	xs := t.boundaryXs()
	i := t.runeAtX(xs, x-(t.Bounds.X+tfPadX())+t.scrollX)
	switch {
	case detail >= 3:
		t.SelectAll()
	case detail == 2:
		s, e := t.wordRangeAt(min(i, max(0, len([]rune(t.Value))-1)))
		t.selStart, t.selEnd, t.anchor = s, e, s
		t.touch()
		t.Invalidate()
	default:
		t.selStart, t.selEnd, t.anchor = i, i, i
		t.touch()
		t.Invalidate()
	}
}

func (t *TextField) OnPointerDrag(x, _ float32) {
	xs := t.boundaryXs()
	i := t.runeAtX(xs, x-(t.Bounds.X+tfPadX())+t.scrollX)
	if t.selStart != t.anchor || t.selEnd != i {
		t.selStart, t.selEnd = t.anchor, i
		t.touch()
		t.Invalidate()
	}
}

func clampInt(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

File: src/ui/textfield.ts
import { withAlpha } from '../gfx/color';
import { font, MONO } from '../gfx/font';
import type { Font } from '../gfx/font';
import { inset, rect } from '../gfx/geom';
import type { DisplayList } from '../gfx/painter';
import type { TextRun } from '../gfx/text/shaper';
import { metrics } from './metrics';
import type { SemanticSpec } from './semantics';
import { funnel } from './textinput';
import { theme } from './theme';
import { paintFocusRing } from './widgets';
import { Size, Widget } from './widget';

const WORD_CH = /[\p{L}\p{N}_]/u;

/**
 * Single-line text input. Editing state (value, selection) is mirrored from
 * the hidden textarea funnel, which owns the keyboard. This widget owns the
 * mouse and all rendering. Selection indices are UTF-16 code units to match
 * the textarea. Glyph positions come from the shaper.
 */
export class TextField extends Widget {
  value = '';
  // per-node radius override (null = the radius.control metric)
  radius: number | null = null;
  placeholder = '';
  // text size in logical px (`size` prop) and family (`mono` prop)
  fontSize = 14;
  mono = false;

  sensitive = false;

  // fires on every edit (wire-level debouncing happens in the protocol layer)
  onInput: ((v: string) => void) | null = null;
  onCommit: ((v: string) => void) | null = null;

  protected selStart = 0;
  protected selEnd = 0;
  protected anchor = 0;
  private scrollX = 0;
  protected caretOn = true;
  private blinkTimer: ReturnType<typeof setInterval> | null = null;
  private lastCommitted = '';

  // last-seen values of the server's one-shot command props
  resetSeq = 0;
  overrideSeq = 0;

  get focused(): boolean {
    return this.ui?.isFocused(this) ?? false;
  }

  // multiline fields (TextArea) let <enter> insert a newline in the funnel
  get multiline(): boolean {
    return false;
  }

  // visual-line caret movement. only meaningful for multiline fields
  verticalMove(_dir: 1 | -1, _extend: boolean): void {}

  protected fieldFont(): Font {
    return font(this.fontSize, this.mono ? { family: MONO } : {});
  }

  override intrinsicSize(): Size {
    return { w: this.width ?? 220, h: this.height ?? Math.max(metrics.controlHeight, Math.round(this.fontSize * 1.7)) };
  }

  override get interactive(): boolean {
    return true;
  }

  override get focusable(): boolean {
    return true;
  }

  override cursor(): string {
    return 'text';
  }

  override onFocusChange(focused: boolean): void {
    if (focused) {
      this.lastCommitted = this.value;
      funnel.acquire(this);
      this.startBlink();
    } else {
      funnel.release(this);
      this.stopBlink();
      this.commit();
    }
    this.invalidate();
  }

  override semantics(): SemanticSpec {
    return { role: 'textbox', label: this.placeholder || 'text field', value: this.value };
  }

  // screenreader activation: take focus so the input funnel engages
  override activate(): void {
    this.ui?.setFocus(this, false);
  }

  // commit points (<enter>, blur) only fire if the value actually changed
  commit(): void {
    if (this.value === this.lastCommitted) return;
    this.lastCommitted = this.value;
    this.onCommit?.(this.value);
  }

  /**
   * Escape-to-revert: back to the value the field had at focus (or last
   * commit). Returns false on a clean field so Escape bubbles onward
   * (dialogs, selection clearing).
   */
  revert(): boolean {
    if (this.value === this.lastCommitted) return false;
    this.value = this.lastCommitted;
    funnel.refresh(this); // resync the hidden textarea, caret to the end
    this.onInput?.(this.value);
    this.invalidate();
    return true;
  }

  /**
   * Server-driven focus (`Focus()` in the SDK). Unlike a plain setFocus this
   * also (re)binds the input funnel: if focus was nominally on this field but
   * the funnel was released (a button click blurred the textarea, devtools
   * stole focus, ...), setFocus alone would no-op and keystrokes would vanish.
   */
  override grabFocus(): void {
    this.ui?.setFocus(this, false);
    if (this.focused) funnel.acquire(this);
  }

  /**
   * Server-forced clear (`ClearValue` in the SDK): unlike a plain value set,
   * this applies even while focused. The sanctioned exception to local echo
   * for submit-and-keep-typing flows. The funnel textarea is resynced so the
   * next keystroke starts from the empty string.
   */
  forceClear(): void {
    this.value = '';
    this.lastCommitted = '';
    this.selStart = this.selEnd = this.anchor = 0;
    this.scrollX = 0;
    funnel.refresh(this);
    this.invalidate();
  }

  /**
   * Server-forced value (`SetValueNow` in the SDK): applies even while
   * focused. The override half of the local-echo contract, for validation
   * and formatting. The pushed value becomes the new baseline (Escape
   * reverts to it, no commit fires unless the user edits again).
   */
  forceValue(v: string): void {
    this.value = v;
    this.lastCommitted = v;
    funnel.refresh(this);
    this.invalidate();
  }

  // called by the funnel after every edit
  syncFromTextarea(value: string, selStart: number, selEnd: number): void {
    const changed = this.value !== value;
    this.value = value;
    this.syncSelection(selStart, selEnd);
    if (changed) this.onInput?.(value);
    this.invalidate();
  }

  // called by the funnel on selection movement
  syncSelection(selStart: number, selEnd: number): void {
    if (this.selStart === selStart && this.selEnd === selEnd) return;
    this.selStart = selStart;
    this.selEnd = selEnd;
    this.caretOn = true; // moving the caret keeps it visible
    this.invalidate();
  }

  override paintSelf(dl: DisplayList): void {
    const b = this.bounds;
    if (this.ui?.showFocusRing(this)) paintFocusRing(dl, b, this.radius ?? metrics.radiusControl);
    dl.rect(b, {
      color: theme.panelInset,
      radius: this.radius ?? metrics.radiusControl,
      borderWidth: this.focused ? metrics.borderWidth * 1.5 : metrics.borderWidth,
      borderColor: this.focused ? theme.accent : theme.controlEdge,
    });

    const run = this.run();
    const innerW = b.w - metrics.spacePad * 2;
    const textY = b.y + (b.h - (run.ascent + run.descent)) / 2;
    const textH = run.ascent + run.descent;
     // keep the caret inside the viewport by sliding the text
    if (this.focused) {
      const caretX = this.xAtCU(this.selEnd);
      if (caretX - this.scrollX > innerW) this.scrollX = caretX - innerW;
      if (caretX - this.scrollX < 0) this.scrollX = caretX;
    }
    const textX = b.x + metrics.spacePad - this.scrollX;
     dl.pushClip(inset(b, 1.5));
    if (this.value === '' && this.placeholder) {
      dl.text(this.placeholder, b.x + metrics.spacePad, textY, this.fieldFont(), theme.inkFaint);
    }
    if (this.focused && this.selStart !== this.selEnd) {
      const x0 = this.xAtCU(Math.min(this.selStart, this.selEnd));
      const x1 = this.xAtCU(Math.max(this.selStart, this.selEnd));
      dl.rect(rect(textX + x0 - 1, textY - 1, x1 - x0 + 2, textH + 2), {
        color: withAlpha(theme.accent, 0.35),
        radius: 2,
      });
    }
    dl.text(this.sensitive ? '*'.repeat(this.value.length) : this.value, textX, textY, this.fieldFont(), theme.ink);
    if (this.focused && this.selStart === this.selEnd && this.caretOn) {
      dl.rect(rect(textX + this.xAtCU(this.selEnd), textY - 1, 1.5, textH + 2), {
        color: theme.accent,
      });
    }
    dl.popClip();
}
 override onPointerDown(x: number, _y: number, detail: number): void {
    const cu = this.cuAtX(x - (this.bounds.x + metrics.spacePad) + this.scrollX);
    if (detail >= 3) {
      funnel.setSelection(0, this.value.length);
      this.anchor = 0;
    } else if (detail === 2) {
      const [s, e] = this.wordRangeCU(cu);
      funnel.setSelection(s, e);
      this.anchor = s;
    } else {
      funnel.setSelection(cu, cu);
      this.anchor = cu;
    }
  }

  override onPointerDrag(x: number, _y: number): void {
    const cu = this.cuAtX(x - (this.bounds.x + metrics.spacePad) + this.scrollX);
    if (cu >= this.anchor) funnel.setSelection(this.anchor, cu);
    else funnel.setSelection(cu, this.anchor, 'backward');
  }

  // -- geometry helpers ---------------------------------------------------------

  private run(): TextRun {
      return this.ui!.measure(this.fieldFont(), this.sensitive? '*'.repeat(this.value.length) : this.value);
  }

  // X offset (from the text origin) of a UTF-16 boundary
  private xAtCU(cu: number): number {
    const run = this.run();
    let units = 0;
    for (const g of run.glyphs) {
      if (units >= cu) return g.x;
      units += g.ch.length;
    }
    return run.width;
  }

  // nearest UTF-16 boundary to an x offset from the text origin
  private cuAtX(x: number): number {
    const run = this.run();
    let best = 0;
    let bestD = Infinity;
    let units = 0;
    const consider = (bx: number, cu: number) => {
      const d = Math.abs(x - bx);
      if (d < bestD) {
        bestD = d;
        best = cu;
      }
    };
    for (const g of run.glyphs) {
      consider(g.x, units);
      units += g.ch.length;
    }
    consider(run.width, units);
    return best;
  }

  protected wordRangeCU(cu: number): [number, number] {
    const cps: { ch: string; off: number }[] = [];
    let off = 0;
    for (const ch of this.value) {
      cps.push({ ch, off });
      off += ch.length;
    }
    if (!cps.length) return [0, 0];
    let i = cps.findIndex((c) => c.off + c.ch.length > cu);
    if (i < 0) i = cps.length - 1;
    const isWord = (s: string) => WORD_CH.test(s);
    const target = isWord(cps[i]!.ch);
    let s = i;
    let e = i + 1;
    while (s > 0 && isWord(cps[s - 1]!.ch) === target) s--;
    while (e < cps.length && isWord(cps[e]!.ch) === target) e++;
    return [cps[s]!.off, e < cps.length ? cps[e]!.off : this.value.length];
  }

  private startBlink(): void {
    this.stopBlink();
    this.caretOn = true;
    this.blinkTimer = setInterval(() => {
      this.caretOn = !this.caretOn;
      this.invalidate();
    }, 530);
  }

  private stopBlink(): void {
    if (this.blinkTimer != null) {
      clearInterval(this.blinkTimer);
      this.blinkTimer = null;
    }
  }
}
